### PR TITLE
`FreeTimeFinder` - algorithm to find a date that suits all required users and most non required

### PR DESCRIPTION
**Pull request by sereczeq**
Don't check the other files yet. They're done by Mustafa. **Please only check `FreeTimeFinder`**

### CODE REVIEW

**Review by Kaspr-whtev — CHANGES_REQUESTED** [review]

File: backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java [hunk 1]
```diff
@@ -40,34 +42,110 @@ public String getRefreshToken() {
...
+     * @see GoogleHelper#getFreeBusyIntervalForAll(Date, Date, Map<String, Map<Long, Set<String>>>)
+     * @see GoogleHelper#getFreeIntervalForAll(Map<String, Map<Long, List<Interval>>>, Date, Date, Duration, List<Entity_PresetAvailability>)
+     * @see Converter#convertIntervalMapToListOfSchedulingInfo(Map<String, Map<Long, List<Interval>>>)
+     * @see FreeTimeFinder#getBestStartDate(List<SchedulingInfo>, Duration)
+     */
+    public DateTime getStartDate(DTO_NewEventDetail newEventDetail, Map<String, Map<Long, Set<String>>> refreshTokenForAllGuests) throws ExecutionException, InterruptedException {
+        Duration durationInMinutes = Duration.standardMinutes(newEventDetail.duration());
+        Map<String, Map<Long, List<Interval>>> freeBusyIntervalsForAllUsers = getFreeBusyIntervalForAll(newEventDetail.start_date(), newEventDetail.end_date(), refreshTokenForAllGuests);
+        Map<String, Map<Long, List<Interval>>> freeIntervalsForAllUsers = getFreeIntervalForAll(freeBusyIntervalsForAllUsers, newEventDetail.start_date(), newEventDetail.end_date(), durationInMinutes, newEventDetail.event_preset_detail().preset_availability());
```
Comment (sereczeq) [inline backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java:78]: Make a class or a struct to hold this data

File: backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java [hunk 2]
```diff
@@ -40,34 +42,110 @@ public String getRefreshToken() {
...
+        List<SchedulingInfo> schedulingInfos = Converter.convertIntervalMapToListOfSchedulingInfo(freeIntervalsForAllUsers);
+        Date startTime = FreeTimeFinder.getBestStartDate(schedulingInfos, durationInMinutes);
+        assert startTime != null;
```
Comment (sereczeq) [inline backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java:81]: [Assertions should never be triggered in production code, and are indicative of a bug or misuse of a code path](https://stackoverflow.com/questions/2758224/what-does-the-java-assert-keyword-do-and-when-should-it-be-used)

File: backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java [hunk 3]
```diff
@@ -40,34 +42,110 @@ public String getRefreshToken() {
...
+     * @param startDate the date after which an event should be scheduled 
+     * @param endDate the date before which an event should be scheduled. 
+     * @param usersRefreshToken a map containing the obligation to attend for all registered guests and the refresh token for all of their google accounts. The map structure is the following: Map<Obligation to attend, Map<PlanIt User Id, Set<Refresh Token>>>
+     * @return Map<String, Map<Long, List<Interval>>> a map containing all FreeBusy intervals mapped to their PlanIt User Id, which mapped to the obligation of attending the event. Map<Obligation to attend, Map<PlanIt User Id, List<FreeBusy Intervals>>>
+     * @see GoogleHelper#getFreeBusy
+     * @see Converter#convertFutureMapToIntervalMap
+     * @see CompletableFuture
+     */
+    private Map<String, Map<Long, List<Interval>>> getFreeBusyIntervalForAll(Date startDate, Date endDate, Map<String, Map<Long, Set<String>>> usersRefreshToken) throws ExecutionException, InterruptedException {
+        Map<String, Map<Long, List<CompletableFuture<List<Interval>>>>> futureCallsMap = new HashMap<>();
```
Comment (sereczeq) [inline backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java:98]: This is unreadable

File: backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java [hunk 4]
```diff
@@ -40,34 +42,110 @@ public String getRefreshToken() {
...
+                for (String refreshToken : usersRefreshToken.get(obligation).get(userId).stream().toList()) {
+                    size += 1;
+                    CompletableFuture<List<Interval>> futureCall = CompletableFuture.supplyAsync(() -> {
+                        try {
+                            return getFreeBusy(startDate, endDate, refreshToken);
+                        } catch (IOException e) {
+                            throw new RuntimeException(e);
+                        }
+                    });
+                    futureCalls.add(futureCall);
```
Comment (sereczeq) [inline backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java:115]: Why do you keep `size` variable. As far as I understand each time `size` is incremented you also add elements to `futureCalls` and `futureIntervalForAUser` so size can be infered from that

File: backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java [hunk 5]
```diff
@@ -76,5 +154,60 @@ private List<Interval> getBusyIntervals(FreeBusyResponse freeBusyResponse) {
...
+        Map<String, Map<Long, List<Interval>>> result = new HashMap<>();
+        List<CompletableFuture<List<Interval>>> futureCalls = new ArrayList<>();
+        int size = -1;
+        for (String key : freeBusyForAll.keySet()) {
+            freeIntervalsForAllUsers.put(key, new HashMap<>());
+            for (Long id : freeBusyForAll.get(key).keySet()) {
+                size += 1;
+                CompletableFuture<List<Interval>> futureCall = CompletableFuture.supplyAsync(() -> getFreeIntervalsForAUser(freeBusyForAll.get(key).get(id), startDate, endDate, duration, presetAvailabilities));
+                freeIntervalsForAllUsers.get(key).put(id, futureCall);
+                futureCalls.add(futureCall);
```
Comment (sereczeq) [inline backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java:180]: similar as above, size seems redundant here

File: backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java [hunk 6]
```diff
@@ -76,5 +154,60 @@ private List<Interval> getBusyIntervals(FreeBusyResponse freeBusyResponse) {
...
+            }
+        }
+        CompletableFuture<Void> combinedCalls = CompletableFuture.allOf(futureCalls.toArray(new CompletableFuture[size]));
+        combinedCalls.get();
+
+        for (String key : freeIntervalsForAllUsers.keySet()) {
+            result.put(key, new HashMap<>());
+            for (Long id : freeIntervalsForAllUsers.get(key).keySet()) {
+                result.get(key).put(id, freeIntervalsForAllUsers.get(key).get(id).get());
```
Comment (sereczeq) [inline backend/src/main/java/planit/people/preparation/GoogleConnector/GoogleHelper.java:189]: bleeeh

**Review by sereczeq — COMMENTED** [review]
I'd change some things, but overall good. Love the documentation

File: backend/src/main/java/planit/people/preparation/Scheduling/Converter.java [hunk 7]
```diff
@@ -26,4 +44,54 @@ public static String decodeURLString(String url) {
...
+     * 
+     * @param freeIntervalsForAllUsers a map containing all free intervals mapped to their PlanIt User Id, which is mapped to the obligation to attend the event. Map<Obligation to attend, Map<PlanIt User Id, List<Free Time Slots Intervals>>> 
+     * @return List<SchedulingInfo> a list of SchedulingInfo objects. 
+     * @see SchedulingInfo 
+     */
+    public static List<SchedulingInfo> convertIntervalMapToListOfSchedulingInfo(Map<String, Map<Long, List<Interval>>> freeIntervalsForAllUsers) {
+        List<SchedulingInfo> result = new ArrayList<>();
+        for (String key : freeIntervalsForAllUsers.keySet()) {
+            for (Long id : freeIntervalsForAllUsers.get(key).keySet()) {
+                if (key.equals("required")) {
```
Comment (sereczeq) [inline backend/src/main/java/planit/people/preparation/Scheduling/Converter.java:88]: Don't hardcode strings in the middle of the code

File: backend/src/main/java/planit/people/preparation/Scheduling/Scheduler.java [hunk 8]
```diff
@@ -193,25 +207,99 @@ private static Interval merge(Interval first, Interval second) {
...
+     * @see Schedular#splitInterval(Interval, Duration)
      */
-    private static Interval getFirstIntervalMatchingDuration(List<Interval> available, Duration duration) {
+    private static List<Interval> filterIntervalsToMatchDuration(List<Interval> available, Duration duration) {
+        var filtered = new ArrayList<Interval>();
         for (Interval interval : available) {
-            Duration intervalDuration = interval.toDuration();
-            if (intervalDuration.compareTo(duration) >= 0) {
-                return new Interval(interval.getStart(), duration);
+            filtered.addAll(splitInterval(interval, duration));
```
Comment (sereczeq) [inline backend/src/main/java/planit/people/preparation/Scheduling/Scheduler.java:219]: I think just `filtered.add()` would work here, no need for `addAll()`

File: backend/src/main/java/planit/people/preparation/Scheduling/Scheduler.java [hunk 9]
```diff
@@ -193,25 +207,99 @@ private static Interval merge(Interval first, Interval second) {
...
+     * @param interval the interval to be sliced
+     * @param duration the duration to be used to determine the size of each slice. 
+     * @return List<Interval> a list of sliced intervals. 
+     */
+    private static List<Interval> splitInterval(Interval interval, Duration duration) {
+        DateTime startTime = interval.getStart();
+        DateTime endTime = interval.getEnd();
+        List<Interval> result = new ArrayList<>();
+        while (true) {
+            if (interval.toDuration().compareTo(duration) >= 0) {
```
Comment (sereczeq) [inline backend/src/main/java/planit/people/preparation/Scheduling/Scheduler.java:295]: `while true` but first line is an if check. Put that condition inside while

File: backend/src/main/java/planit/people/preparation/Scheduling/Scheduler.java (hunk 8 above)
Comment (sereczeq) [inline backend/src/main/java/planit/people/preparation/Scheduling/Scheduler.java:219]: This <b> functionality is wrong </b> This method is supposed to filter out intervals shorter than duration, not cut them. Slicing the intervals will create a lot of false negatives while searching for the time slots

**Review by sereczeq — COMMENTED** [review]
This cannot be merged! Changes break the functionality of the code

**Review by Kartm — CHANGES_REQUESTED** [review]

**Review by Kartm — APPROVED** [review]

**Pull request merged**